torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,715 @@
+
+  NCCL_CHECK(ncclGroupStart());
+  for (size_t i = 0; i < data.size(); ++i) {
+
+    gpuGuard.setDevice(data[i].get_device());
+    auto stream = THCState_getCurrentStream(THDGetCudaState());
+
+    NCCL_CHECK(ncclReduce(data[i].data_ptr(),
+                          data[i].data_ptr(),
+                          data[i].numel(),
+                          _getNcclDataType(data[i].type().scalarType()),
+                          ncclOp[operation],
+                          dstRank * data.size(),
+                          (*comms)[i],
+                          stream));
+    THCudaCheck(cudaEventRecord((*events)[i], stream));
+  }
+  NCCL_CHECK(ncclGroupEnd());
+
+  cudaFreeMutexLock.unlock();
+}
+
+void DataChannelNccl::reduce(at::Tensor& data,
+                             THDReduceOp operation,
+                             rank_type dstRank,
+                             THDGroup groupId) {
+
+  std::vector<at::Tensor> dataVec = {data};
+  reduce(dataVec, operation, dstRank, groupId);
+}
+
+void DataChannelNccl::broadcast(std::vector<at::Tensor>& data,
+                                rank_type srcRank,
+                                THDGroup groupId) {
+
+  std::unique_lock<std::mutex> channelLock(_mutex);
+
+  // Check the tensor vector for consistency
+  if (!_tensorCheckHelper(data, data)) {
+    return;
+  }
+  _checkGroupIdValid(groupId);
+
+  auto ncclResourcePair = _getNcclResourcePair(data, groupId);
+  auto comms = ncclResourcePair.first;
+  auto events = ncclResourcePair.second;
+
+  // Guard GPU device
+  AutoGPU gpuGuard;
+
+  std::unique_lock<std::mutex> cudaFreeMutexLock(
+      *(THCCachingAllocator_getCudaFreeMutex()));
+
+  NCCL_CHECK(ncclGroupStart());
+  for (size_t i = 0; i < data.size(); ++i) {
+
+    gpuGuard.setDevice(data[i].get_device());
+    auto stream = THCState_getCurrentStream(THDGetCudaState());
+
+    NCCL_CHECK(ncclBcast(data[i].data_ptr(),
+                         data[i].numel(),
+                         _getNcclDataType(data[i].type().scalarType()),
+                         srcRank * data.size(),
+                         (*comms)[i],
+                         stream));
+    THCudaCheck(cudaEventRecord((*events)[i], stream));
+  }
+  NCCL_CHECK(ncclGroupEnd());
+
+  cudaFreeMutexLock.unlock();
+}
+
+
+void DataChannelNccl::broadcast(at::Tensor& data,
+                                rank_type srcRank,
+                                THDGroup groupId) {
+
+  std::vector<at::Tensor> dataVec = {data};
+  broadcast(dataVec, srcRank, groupId);
+}
+
+
+void DataChannelNccl::barrier(THDGroup groupId) {
+  throw std::runtime_error("DataChannelNccl does not support barrier");
+}
+
+
+THDGroup DataChannelNccl::newGroup(const std::vector<rank_type>& ranks) {
+
+  std::unique_lock<std::mutex> channelLock(_mutex);
+
+  auto newGroup = DataChannel::Group(ranks, _numProcesses - 1);
+  THDGroup newGroupId = static_cast<THDGroup>(_groups.size());
+
+  // Insert the current group
+  _groups.insert({
```

Comment:
It seems that you're not using `_groups` for anything useful. Why even bother with maintaining them?

Reply:
I guess? Keeping a single counter would be a lot simpler than maintaining a whole map that's unused.